idaes/core/base/property_base.py

```diff
@@ -504,6 +505,47 @@ def report(self, index=(0), true_state=False, dof=False, ostream=None, prefix=""
 
         ostream.write("\n" + "=" * max_str_length + "\n")
 
+    def build_port(self, target_block, port_name, doc=None, subset=None):
+        """
+        Constructs a Port based on this StateBlock attached to the target block.
+
+        Args:
+            target_block - block to which Port should be attached
+            port_name - name to use for Port
+            doc - doc string or Prot object
+            subset - slicer representing a subset of indices of StateBlock which
+                should be included in Port
+
+        Returns:
+            Port object
+        """
+        if subset is None:
+            subset = self[...]
+
+        # Create empty Port
+        p = Port(doc=doc)
+        setattr(target_block, port_name, p)
+
+        # Get dict of Port members and names
+        # Need to get a representative member of StateBlockDatas
+        i0 = self.index_set().first()
+        member_list = self[i0].define_port_members()
+
+        # Create References for port members
+        for s in member_list:
+            if not member_list[s].is_indexed():
+                slicer = subset.component(member_list[s].local_name)
+            else:
+                slicer = subset.component(member_list[s].local_name)[...]
+
+            r = Reference(slicer)
+            setattr(target_block, "_" + s + "_" + port_name + "_ref", r)
```

Comment:
Should there be a method to get this "port reference" name from the component name (`s`)? This would be useful if we every needed to programmatically access these references by name.

Reply:
We could do that, but where would that method live? Also, the user would have to be able to provide the target block, component name and port name in the method call, meaning they basically know the Reference name anyway.

Reply:
Good question. It probably would make more sense for this method to live in `unit_model.py` rather than `property_base.py`. Something like:
```python
def get_port_reference_name(component_name, port_name):
    return "_%s_%s_ref" % (component_name, port_name)
```
Then the `add_port` method could do something like:
```python
port, name_ref_list = state_block.build_port(self, name, doc)
self.add_component(name, port)
for comp_name, ref in name_ref_list:
    ref_name = get_port_reference_name(comp_name, name)
    self.add_component(ref_name, ref)
```
Of course, a unit model could override this with their own `add_port` if they want different names, and the `get_port_reference_name` function could be imported by anybody who wants to identify a port reference from a particular state variable.